pq-crypto/sike_r3/sikep434r3_kem.c

```diff
@@ -0,0 +1,127 @@
+    shake256(h, SIKE_P434_R3_MSG_BYTES, jinvariant, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] = temp[i] ^ h[i];
+    }
+
+    /* Generate shared secret ss <- H(m||ct) */
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], ct, SIKE_P434_R3_CIPHERTEXT_BYTES);
+    shake256(ss, SIKE_P434_R3_SHARED_SECRET_BYTES, temp, SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES);
+
+    return EVP_PQ_SUCCESS;
+}
+
+/* SIKE's decapsulation
+ * Input:   secret key sk         (SIKE_P434_R3_SECRET_KEY_BYTES = SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES + SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+ *          ciphertext message ct (SIKE_P434_R3_CIPHERTEXT_BYTES = SIKE_P434_R3_PUBLIC_KEY_BYTES + SIKE_P434_R3_MSG_BYTES bytes)
+ * Outputs: shared secret ss      (SIKE_P434_R3_SHARED_SECRET_BYTES bytes) */
+int evp_sike_p434_r3_crypto_kem_dec(unsigned char *ss, const unsigned char *ct, const unsigned char *sk)
+{
+    //POSIX_ENSURE(s2n_pq_is_enabled(), S2N_ERR_PQ_DISABLED);
+
+    unsigned char ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES];
+    unsigned char jinvariant_[SIKE_P434_R3_FP2_ENCODED_BYTES];
+    unsigned char h_[SIKE_P434_R3_MSG_BYTES];
+    unsigned char c0_[SIKE_P434_R3_PUBLIC_KEY_BYTES];
+    unsigned char temp[SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES];
+
+    /* Decrypt */
+    EphemeralSecretAgreement_B(sk + SIKE_P434_R3_MSG_BYTES, ct, jinvariant_);
+    shake256(h_, SIKE_P434_R3_MSG_BYTES, jinvariant_, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        temp[i] = ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] ^ h_[i];
+    }
+
+    /* Generate ephemeralsk_ <- G(m||pk) mod oA */
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], &sk[SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES], SIKE_P434_R3_PUBLIC_KEY_BYTES);
+    shake256(ephemeralsk_, SIKE_P434_R3_SECRETKEY_A_BYTES, temp, SIKE_P434_R3_PUBLIC_KEY_BYTES+SIKE_P434_R3_MSG_BYTES);
+    ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES - 1] &= SIKE_P434_R3_MASK_ALICE;
+
+    /* Generate shared secret ss <- H(m||ct), or output ss <- H(s||ct) in case of ct verification failure */
+    EphemeralKeyGeneration_A(ephemeralsk_, c0_);
+
+    /* Verify ciphertext.
+     * If c0_ and ct are NOT equal, decaps failed and we overwrite the shared secret
+     * with pseudorandom noise (ss = H(s||ct)) by performing the copy (dont_copy = false).
+     *
+     * If c0_ and ct are equal, then decaps succeeded and we skip the overwrite and output
+     * the actual shared secret: ss = H(m||ct) (dont_copy = true). */
+    //bool dont_copy = s2n_constant_time_equals(c0_, ct, S2N_SIKE_P434_R3_PUBLIC_KEY_BYTES);
+    // POSIX_GUARD(s2n_constant_time_copy_or_dont(temp, sk, S2N_SIKE_P434_R3_MSG_BYTES, dont_copy));
+    int dont_copy = CRYPTO_memcmp(c0_, ct, SIKE_P434_R3_PUBLIC_KEY_BYTES);
+    if (dont_copy == 0) {
+        OPENSSL_memcpy(temp, sk, SIKE_P434_R3_MSG_BYTES);
```

Comment:
Yeah - this doesn't really do much more than verify that the length is non-zero before performing a standard `memcpy`.